### DISCUSSION TITLE
Adding conditional input to `jft.Model` or `jft.WrappedCall`

### DISCUSSION THREAD

**Discussion by homerjed** [NIFTy-PPL/NIFTy#11] [Q&A]
I've checked the examples but I don't think there's one that answers this question.

I have a model that draws a signal `s` from `p(s|y, xi)` where `xi` are some latent variables and `y` is a conditioning variable, e.g. some continuous number. 

In a case without the conditioning variable `y` my model takes a `xi` draw and creates a signal `s`  using a `jft.WrappedCall` as 

```python
signal = jft.WrappedCall(
    sample_fn,  # function with one arg, xi, such that s = sample_fn(xi)
    name="xi", 
    shape=xi_shape,
    dtype=jnp.float64,
    white_init=True
)

# Diagonal noise covariance
noise_cov = lambda x: 2. ** 2. * x
noise_cov_inv = lambda x: 2. ** -2. * x

key_noise, key_xi, key_init = jr.split(key, 3)

# Create synthetic data
sigma = noise_cov(jft.ones_like(signal.target)) ** 0.5
noise_truth = sigma * jft.random_like(key_noise, signal.target)
xi_truth = jft.random_like(key_xi, signal.domain) 
signal_truth = signal(xi_truth)

data = signal_truth + noise_truth

nll = jft.Gaussian(data, noise_cov_inv).amend(signal)

# ... run MGVI
```

so what is a way to incorporate a conditioning input `y`? 

I may want to marginalise over the `xi` I fit for in my posterior, so I guess they should be distinct somehow (e.g. as opposed to simply splitting off a condition variable from the `xi` input).

I assume it would be better to use something like

```python
class Forward(jft.Model):
  def __init__(self, correlated_field):
    self._cf = correlated_field
    # Track a method with which a random input for the model. This is not
    # strictly required but is usually handy when building deep models.
    super().__init__(init=correlated_field.init)

  def __call__(self, x):
    # NOTE, any kind of masking of the output, non-linear and linear
    # transformation could be carried out here. Models can also combined and
    # nested in any way and form.
    return jnp.exp(self._cf(x))
```

as on the repo landing page, but i'm not sure how to write this out. 

Cheers

**Comment by ph-frank** [comment]
Hi there,
Thanks for the question and your interest in NIFTy!

Let me expand on this one a little as the exact way to go about depends on the situation:

First, with the switch to `jax`, building models in nifty became fully functional. This means that we may directly invoke python functions as forward models in e.g. likelihoods via: `likelihood.amend(sample_fn)`. The forward model `sample_fn` is expected to be a [pure function](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions) and takes a single input `primals`.
Furthermore, `primals` can be an arbitrary [pytree](https://jax.readthedocs.io/en/latest/pytrees.html#pytrees) (e.g. a list, tuple, dict of numerical input), thus how inputs are organized in models and what role they play is left to the user.

So in order to only achieve conditioning, something as minimal as:
```
y = # conditional value

llh = jft.Gaussian(data, noise_cov_inv)
# Amend forward model with partially inserted conditional y
conditional_llh = llh.amend(partial(sample_fn, y=y))

# Explicitly initialize random position instead of invoking llh.init
pos_init = jax.random.normal(key, xi_shape)

... # run vi
```

Alternatively one may also build the likelihood without conditioning, and condition later using the likelihood's builtin method `Likelihood.freeze` which constructs a new likelihood with partially inserted `primals` as input.

That being said, instead of working with functions directly, it may be convenient to work with a nifty.re `Model` (note that `WrappedCall` is just a helper function to create a `Model`) in case one wants to associate additional information with a function/forward model (such as e.g. shape/dtype of the input or how to initialize it). Whether or not that is intended here depends on the detailed circumstances of the problem.

Hope this helps! Cheers
